torch/csrc/jit/graph_executor.cpp

```diff
@@ -409,70 +412,117 @@ struct GraphExecutorImpl {
     return false;
   }
 
-
-  // remove ReplaceIfUndef(v, replacement) nodes that consume inputs with 'v' if
-  // the input is defined, and 'replacement' if it is not.
-  // Note: this is a very limited pass. It looks at undefined inputs,
-  // and cleans up ReplaceIfUndef nodes inserted by autodiff.
+  // propagate undefined information through a gradient graph
+  // Note: this is a very limited pass. It only propagates undefines for
+  // operations generated by the symbolic autodiff code and cleans up
+  // AutogradAdds when possible. Outputs of other nodes are conservatively
+  // marked MaybeDefined and not optimized.
   void specializeUndef(Graph & g, const ArgumentSpec & spec) {
+    enum class State { Defined, Undefined, MaybeDefined };
+    std::unordered_map<Value*, State> state;
     for(size_t i = 0; i < spec.size(); i++) {
-      std::vector<Value*> to_replace;
-      // do not edit in place, since it invalidates uses iterator
-      for(auto u : g.inputs()[i]->uses()) {
-        if(u.user->kind() == prim::ReplaceIfUndef) {
-          to_replace.push_back(u.user->output());
-        }
-      }
-      for(auto v : to_replace) {
-        // if it is defined, then we replace with 'v' if not,
-        // we replace with 'replacement' which is normally just a zero tensor
-        int idx = spec.tensorInfo(i).defined() ? 0 : 1;
-        v->replaceAllUsesWith(v->node()->inputs()[idx]);
-        v->node()->destroy();
-      }
+      state[g.inputs()[i]] = spec.tensorInfo(i).defined() ? State::Defined : State::Undefined;
     }
-  }
-  // a + 0 -> a
-  // 0 + a -> a
-  void propagateZeros(Graph & g) {
     for(auto it = g.nodes().begin(); it != g.nodes().end(); ++it) {
-      if(it->kind() == aten::add && it->inputs().size() == 2 && at::Scalar(it->t(attr::alpha)).toDouble() == 1.0) {
-        if(isZero(it->inputs()[0])) {
-          it->output()->replaceAllUsesWith(it->inputs()[1]);
-          it.destroyCurrent();
-        } else if(isZero(it->inputs()[1])) {
-          it->output()->replaceAllUsesWith(it->inputs()[0]);
-          it.destroyCurrent();
-        }
+      auto n = *it;
+      switch(n->kind()) {
+        case prim::LinearBlock: {
+          auto all_undefined =
+              std::all_of(n->inputs().begin(), n->inputs().end(), [&](Value* v) {
+                return state[v] == State::Undefined;
+              });
+          // Property 1: if all the gradInputs to the LinearBlock are undefined
+          // then the gradOutputs are also zero and will be represented as undefined nodes
```

Comment:
In particular, these are not *multilinear maps*, because we are not individually linear in each argument.